cmd/agent/run.go

```diff
@@ -13,15 +14,29 @@ import (
 func run(ctx context.Context, options *config.AgentOptions) error {
 	logrus.Infof("Starting VM DHCP Agent: %s", name)
 
-	agent := agent.NewAgent(ctx, options)
+	agent := agent.NewAgent(options)
 
 	httpServerOptions := config.HTTPServerOptions{
 		DebugMode:     enableCacheDumpAPI,
 		DHCPAllocator: agent.DHCPAllocator,
 	}
 	s := server.NewHTTPServer(&httpServerOptions)
 	s.RegisterAgentHandlers()
-	go s.Run()
 
-	return agent.Run()
+	eg, egctx := errgroup.WithContext(ctx)
+
+	eg.Go(func() error {
+		return s.Run()
+	})
+
+	eg.Go(func() error {
+		return agent.Run(egctx)
+	})
+
+	eg.Go(func() error {
+		<-egctx.Done()
+		return s.Stop(egctx)
+	})
+
+	return eg.Wait()
```

Comment:
BTW, if we try to wait `egctx.Done()` in side eg.Go, it might encounter unexpected error like this.

The `group.Wait()` waits until cancel is called in 5 seconds. That means if it doesn't call cancel, it will wait forever.  It might be potential issue here.

```go
func main() {
	ctx, cancel := context.WithCancel(context.Background())
	time.AfterFunc(5*time.Second, cancel)

	group, egctx := errgroup.WithContext(ctx)

	group.Go(func() error {
		return nil
	})

	group.Go(func() error {
		<-egctx.Done()
		fmt.Println("do something")
		return nil
	})

	if err := group.Wait(); err != nil {
		log.Println("error is ", err)
	}

	log.Println("done")
}
```